File: src/dqrobotics/interfaces/coppeliasim/DQ_SerialCoppeliaSimRobot.cpp
/**
(C) Copyright 2024 DQ Robotics Developers

This file is based on DQ Robotics.

    DQ Robotics is free software: you can redistribute it and/or modify
    it under the terms of the GNU Lesser General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    DQ Robotics is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public License
    along with DQ Robotics.  If not, see <http://www.gnu.org/licenses/>.

Contributors:
- Juan Jose Quiroz Omana
       - Responsible for the original implementation.
         The DQ_SerialCoppeliaSimRobot class is partially based on the DQ_SerialVrepRobot class
         (https://github.com/dqrobotics/cpp-interface-vrep/blob/master/include/dqrobotics/interfaces/vrep/DQ_SerialVrepRobot.h)

*/

#include <dqrobotics/interfaces/coppeliasim/DQ_SerialCoppeliaSimRobot.h>


namespace DQ_robotics
{

/**
 * @brief DQ_SerialCoppeliaSimRobot::_initialize_jointnames_from_coppeliasim
 */
void DQ_SerialCoppeliaSimRobot::_initialize_jointnames_from_coppeliasim()
{
    jointnames_ = _get_interface_sptr()->get_jointnames_from_base_objectname(robot_name_);
    base_frame_name_ = jointnames_.at(0);
}


/**
 * @brief DQ_SerialCoppeliaSimRobot::DQ_SerialCoppeliaSimRobot
 * @param robot_name
 * @param coppeliasim_interface_sptr
 */
DQ_SerialCoppeliaSimRobot::DQ_SerialCoppeliaSimRobot(const std::string &robot_name,
                                                     const std::shared_ptr<DQ_CoppeliaSimInterface> &coppeliasim_interface_sptr)
    :DQ_CoppeliaSimRobot(robot_name, coppeliasim_interface_sptr)
{
    _initialize_jointnames_from_coppeliasim();
}


/**
 * @brief DQ_SerialCoppeliaSimRobot::get_joint_names
 * @return
 */
std::vector<std::string> DQ_SerialCoppeliaSimRobot::get_joint_names()
{
    return jointnames_;
}


/**
 * @brief DQ_SerialCoppeliaSimRobot::set_configuration_space_positions
 * @param q
 */
void DQ_SerialCoppeliaSimRobot::set_configuration_space_positions(const VectorXd &q)
{
    _get_interface_sptr()->set_joint_positions(jointnames_,q);
}


/**
 * @brief DQ_SerialCoppeliaSimRobot::get_configuration_space_positions
 * @return
 */
VectorXd DQ_SerialCoppeliaSimRobot::get_configuration_space_positions()
{
    return  _get_interface_sptr()->get_joint_positions(jointnames_);
}


/**
 * @brief DQ_SerialCoppeliaSimRobot::set_target_configuration_space_positions
 * @param q_target
 */
void DQ_SerialCoppeliaSimRobot::set_target_configuration_space_positions(const VectorXd &q_target)
{
    return _get_interface_sptr()->set_joint_target_positions(jointnames_, q_target);
}


/**
 * @brief DQ_SerialCoppeliaSimRobot::get_configuration_space_velocities
 * @return
 */
VectorXd DQ_SerialCoppeliaSimRobot::get_configuration_space_velocities()
{
    return _get_interface_sptr()->get_joint_velocities(jointnames_);
}

/**
 * @brief DQ_SerialCoppeliaSimRobot::set_target_configuration_space_velocities
 * @param v_target
 */
void DQ_SerialCoppeliaSimRobot::set_target_configuration_space_velocities(const VectorXd &v_target)
{
    _get_interface_sptr()->set_joint_target_velocities(jointnames_, v_target);
}

/**
 * @brief DQ_SerialCoppeliaSimRobot::set_configuration_space_torques
 * @param torques
 */
void DQ_SerialCoppeliaSimRobot::set_configuration_space_torques(const VectorXd &torques)
{
    _get_interface_sptr()->set_joint_torques(jointnames_, torques);
}

/**
 * @brief DQ_SerialCoppeliaSimRobot::get_configuration_space_torques
 * @return
 */
VectorXd DQ_SerialCoppeliaSimRobot::get_configuration_space_torques()
{
    return _get_interface_sptr()->get_joint_torques(jointnames_);
}

}

File: matlab/interfaces/coppeliasim/DQ_SerialCoppeliaSimRobot.m
classdef DQ_SerialCoppeliaSimRobot < DQ_CoppeliaSimRobot

    properties (Access = protected)
        jointnames_        cell
        base_frame_name_   string
        joint_control_mode_ DQ_CoppeliaSimInterface_JOINT_CONTROL_MODE 
        robot_is_used_as_visualization_tool_ logical
    end

    methods (Access = protected)
        function obj = DQ_SerialCoppeliaSimRobot(robot_name, coppeliasim_interface_sptr)
            obj = obj@DQ_CoppeliaSimRobot(robot_name, coppeliasim_interface_sptr);
         obj.initialize_jointnames_from_coppeliasim_();
         % // By Default, the robot is controlled by joint positions with both the dynamic engine
         % // and the stepping mode enabled.
         obj.joint_control_mode_ = DQ_CoppeliaSimInterface_JOINT_CONTROL_MODE.POSITION;
         obj.robot_is_used_as_visualization_tool_ = false;
        end

        function initialize_jointnames_from_coppeliasim_(obj)
            obj.jointnames_ = obj.get_interface_sptr_().get_jointnames_from_base_objectname(obj.robot_name_);
            obj.base_frame_name_ = obj.jointnames_{1};
        end
    end
    
    methods
        function set_operation_modes(obj, joint_mode, joint_control_mode)
             arguments
                 obj
                 joint_mode DQ_CoppeliaSimInterface_JOINT_MODE
                 joint_control_mode DQ_CoppeliaSimInterface_JOINT_CONTROL_MODE
             end
             obj.get_interface_sptr_().set_joint_modes(obj.jointnames_, joint_mode);
             obj.get_interface_sptr_().set_joint_control_modes(obj.jointnames_, joint_control_mode);
        end

        function set_robot_as_visualization_tool(obj)

            obj.get_interface_sptr_().set_stepping_mode(false);
            obj.get_interface_sptr_().enable_dynamics(false);
            obj.get_interface_sptr_().set_joint_modes(obj.jointnames_,...
                                                   DQ_CoppeliaSimInterface_JOINT_MODE.KINEMATIC);
            obj.robot_is_used_as_visualization_tool_ = true;
        end

        function set_joint_control_type(obj, joint_control_mode)
            arguments
                obj 
                joint_control_mode DQ_CoppeliaSimInterface_JOINT_CONTROL_MODE
            end
            obj.joint_control_mode_ = joint_control_mode;
            obj.get_interface_sptr_().enable_dynamics(true);
            obj.get_interface_sptr_().set_stepping_mode(true);
            obj.set_operation_modes(DQ_CoppeliaSimInterface_JOINT_MODE.DYNAMIC, obj.joint_control_mode_);
        end

        function set_control_inputs(obj, u)
            if (obj.robot_is_used_as_visualization_tool_)
                    obj.get_interface_sptr_().set_joint_positions(obj.jointnames_, u);
            else
                switch (obj.joint_control_mode_)
                    case DQ_CoppeliaSimInterface_JOINT_CONTROL_MODE.FREE
                    case DQ_CoppeliaSimInterface_JOINT_CONTROL_MODE.FORCE
                    case DQ_CoppeliaSimInterface_JOINT_CONTROL_MODE.VELOCITY
                         obj.get_interface_sptr_().set_joint_target_velocities(obj.jointnames_, u);
                    case DQ_CoppeliaSimInterface_JOINT_CONTROL_MODE.POSITION
                         obj.get_interface_sptr_().set_joint_target_positions(obj.jointnames_, u);           
                    case DQ_CoppeliaSimInterface_JOINT_CONTROL_MODE.SPRING
                    case DQ_CoppeliaSimInterface_JOINT_CONTROL_MODE.CUSTOM
                    case DQ_CoppeliaSimInterface_JOINT_CONTROL_MODE.TORQUE
                        obj.get_interface_sptr_().set_joint_torques(obj.jointnames_, u);
                end
                obj.get_interface_sptr_().trigger_next_simulation_step();
            end
            
        end

        function rtn = get_joint_names(obj)
                rtn = obj.jointnames_;
        end

        function set_configuration_space_positions(obj, q)
            obj.get_interface_sptr_().set_joint_positions(obj.jointnames_,q);
        end

        function q = get_configuration_space_positions(obj)
            q =  obj.get_interface_sptr_().get_joint_positions(obj.jointnames_);
        end 

        function set_target_configuration_space_positions(obj, q_target)
            obj.get_interface_sptr_().set_joint_target_positions(obj.jointnames_, q_target);
        end

        function q_dot = get_configuration_space_velocities(obj)
            q_dot = obj.get_interface_sptr_().get_joint_velocities(obj.jointnames_);
        end

        function set_target_configuration_space_velocities(obj, v_target)
            obj.get_interface_sptr_().set_joint_target_velocities(obj.jointnames_, v_target);
        end

        function set_configuration_space_torques(obj, torques)
            obj.get_interface_sptr_().set_joint_torques(obj.jointnames_, torques);
        end


        function torques = get_configuration_space_torques(obj)
            torques = obj.get_interface_sptr_().get_joint_torques(obj.jointnames_);
        end

    end
end

